Importing the library.

In [1]:
import numpy as np
import pandas as pd

Importing the dataset.

In [2]:
dataset = pd.read_csv("D:\MLprojects\champions.csv")

The dataset is the statistics of all of 147 champions in League of Legends from 27/11/2019 to 27/12/2019.

In [3]:
dataset

Champions  Win Rate  Games Played  KD/A      CS   Gold
0      Aatrox     45.86        164593  1.84  150.57  10395
1        Ahri     51.23        198894  2.60  142.14  10533
2       Akali     45.90        327924  2.12  149.43  10810
3     Alistar     47.76         75860  2.32   33.28   7597
4       Amumu     51.37        123849  2.35  140.60  10266
..        ...       ...           ...   ...     ...    ...
142       Zed     48.85        251931  2.21  153.53  11690
143     Ziggs     49.89         58872  2.39  151.20  10814
144    Zilean     51.84         81726  2.90   64.34   8746
145       Zoe     47.67         83066  2.38  128.90  10254
146      Zyra     51.92        143841  2.34   54.18   9305

[147 rows x 6 columns]

The model will predict the average amount of gold a random champion gained in a game based on Win Rate, Games Played, KD/A, and CS in any time periods.

Divide the dataset into independent variables and dependent variables.

In [4]:
X = dataset.iloc[:, 1:5].values
x = pd.DataFrame(X, columns=["Win Rate", "Games Played", "KD/A", "CS"])

In [5]:
x

Win Rate  Games Played  KD/A      CS
0       45.86      164593.0  1.84  150.57
1       51.23      198894.0  2.60  142.14
2       45.90      327924.0  2.12  149.43
3       47.76       75860.0  2.32   33.28
4       51.37      123849.0  2.35  140.60
..        ...           ...   ...     ...
142     48.85      251931.0  2.21  153.53
143     49.89       58872.0  2.39  151.20
144     51.84       81726.0  2.90   64.34
145     47.67       83066.0  2.38  128.90
146     51.92      143841.0  2.34   54.18

[147 rows x 4 columns]

In [6]:
Y = dataset.iloc[:, 5].values
y = pd.DataFrame(Y, columns=["Gold"])

In [7]:
y

Gold
0    10395
1    10533
2    10810
3     7597
4    10266
..     ...
142  11690
143  10814
144   8746
145  10254
146   9305

[147 rows x 1 columns]

Inspecting which independent variable have the least influence on the dependent variable using backward elimination. Choose the significant level for this model to be SL = 0.05.

In [8]:
import statsmodels.api as sm
x_model = X[: , [0,1,2,3]]
regressor_OLS = sm.OLS(Y,x_model).fit()
regressor_OLS.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                      y   R-squared (uncentered):                   0.989
Model:                            OLS   Adj. R-squared (uncentered):              0.989
Method:                 Least Squares   F-statistic:                              3210.
Date:                Wed, 08 Jan 2020   Prob (F-statistic):                   7.25e-139
Time:                        23:17:53   Log-Likelihood:                         -1239.4
No. Observations:                 147   AIC:                                      2487.
Df Residuals:                     143   BIC:                                      2499.
Df Model:                           4                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
x1           292.7091     13.919     21.030      0.000     265.196     320.222
x2             0.0007      0.001      0.994      0.322      -0.001       0.002
x3         -1821.2366    286.095     -6.366      0.000   -2386.758   -1255.715
x4             0.0724      0.070      1.042      0.299      -0.065       0.210
==============================================================================
Omnibus:                        6.660   Durbin-Watson:                   1.769
Prob(Omnibus):                  0.036   Jarque-Bera (JB):                6.363
Skew:                          -0.498   Prob(JB):                       0.0415
Kurtosis:                       3.214   Cond. No.                     6.70e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 6.7e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

The predictor x2 (Games Played) has the highest p-value, which is 0.322. This p-value is greater than the significant level (SL = 0.05), so remove x2 (Games Played) from the model.

In [9]:
x_model = X[: , [0,2,3]]
regressor_OLS = sm.OLS(Y,x_model).fit()
regressor_OLS.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                      y   R-squared (uncentered):                   0.989
Model:                            OLS   Adj. R-squared (uncentered):              0.989
Method:                 Least Squares   F-statistic:                              4279.
Date:                Wed, 08 Jan 2020   Prob (F-statistic):                   1.67e-140
Time:                        23:17:53   Log-Likelihood:                         -1239.9
No. Observations:                 147   AIC:                                      2486.
Df Residuals:                     144   BIC:                                      2495.
Df Model:                           3                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
x1           296.1103     13.491     21.949      0.000     269.444     322.776
x2         -1837.5002    285.614     -6.434      0.000   -2402.038   -1272.962
x3             0.0700      0.069      1.007      0.316      -0.067       0.207
==============================================================================
Omnibus:                        5.556   Durbin-Watson:                   1.792
Prob(Omnibus):                  0.062   Jarque-Bera (JB):                5.304
Skew:                          -0.462   Prob(JB):                       0.0705
Kurtosis:                       3.107   Cond. No.                     4.18e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 4.18e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

The predictor x3 (CS) has the highest p-value, which is 0.316. This p-value is greater than the significant level (SL = 0.05), so remove x3 (CS) from the model.

In [10]:
x_model = X[: , [0,2]]
regressor_OLS = sm.OLS(Y,x_model).fit()
regressor_OLS.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                      y   R-squared (uncentered):                   0.989
Model:                            OLS   Adj. R-squared (uncentered):              0.989
Method:                 Least Squares   F-statistic:                              6418.
Date:                Wed, 08 Jan 2020   Prob (F-statistic):                   3.05e-142
Time:                        23:17:53   Log-Likelihood:                         -1240.4
No. Observations:                 147   AIC:                                      2485.
Df Residuals:                     145   BIC:                                      2491.
Df Model:                           2                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
x1           296.6543     13.481     22.006      0.000     270.010     323.299
x2         -1841.9561    285.594     -6.450      0.000   -2406.421   -1277.492
==============================================================================
Omnibus:                        5.937   Durbin-Watson:                   1.789
Prob(Omnibus):                  0.051   Jarque-Bera (JB):                5.716
Skew:                          -0.480   Prob(JB):                       0.0574
Kurtosis:                       3.108   Cond. No.                         154.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Two remaining predictors, which are Win Rate and KD/A, have the p-values that are lower than the significant value (SL = 0.05). Moreover, the adjusted R-squared value nearly remains the same (0.989) through the process of backward elimination. Therefore, choose the two independent variables, Win Rate and KD/A, to fit into the model.

In [11]:
X_input = X[: , [0,2]]

In [12]:
X_input

array([[45.86,  1.84],
       [51.23,  2.6 ],
       [45.9 ,  2.12],
       [47.76,  2.32],
       [51.37,  2.35],
       [50.6 ,  2.72],
       [50.31,  2.27],
       [49.17,  2.1 ],
       [51.88,  2.51],
       [52.08,  2.52],
       [44.71,  1.89],
       [50.62,  2.64],
       [51.96,  2.44],
       [52.  ,  2.12],
       [47.96,  2.59],
       [50.88,  2.42],
       [49.42,  1.99],
       [50.61,  2.17],
       [49.12,  2.18],
       [45.98,  2.33],
       [51.39,  2.01],
       [51.72,  2.14],
       [52.09,  2.28],
       [49.64,  2.15],
       [52.1 ,  2.46],
       [50.06,  2.29],
       [47.3 ,  2.33],
       [45.55,  2.42],
       [49.  ,  2.11],
       [50.29,  1.78],
       [53.21,  2.58],
       [49.29,  2.46],
       [47.6 ,  2.27],
       [52.25,  2.28],
       [48.11,  2.03],
       [45.96,  2.22],
       [49.71,  2.3 ],
       [49.62,  2.43],
       [51.89,  2.11],
       [51.09,  1.72],
       [45.63,  1.87],
       [49.27,  2.9 ],
       [52.19,  3.01],
       [51.

In [13]:
x_input = pd.DataFrame(X_input, columns=["Win Rate", "KD/A"])

Splitting the dataset into the Training set and Test set.

In [14]:
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X_input, Y, test_size = 0.2, random_state = 13)

Fitting Multiple Linear Regression to the Training set.

In [15]:
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, Y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

Predicting the Test set result.

In [16]:
Y_pred = regressor.predict(X_test)

Compare the predicted gold values to the real gold values to see the efficiency of the model.

In [17]:
Y_compare = np.array([Y_test, Y_pred]).transpose()
Y_compare_table = pd.DataFrame(Y_compare, columns=["Real Gold Value", "Predicted Gold Value"])

In [18]:
Y_compare_table

Real Gold Value  Predicted Gold Value
0           11036.0          10600.960339
1           10575.0          10173.036212
2           11459.0          10889.907811
3           10814.0          10298.938375
4           10903.0          10749.613035
5           11811.0          10538.650787
6           11097.0          10591.908807
7           11813.0           9351.521584
8            9625.0           9837.154378
9           10290.0           9408.155021
10          11229.0          10203.023363
11           7453.0           6639.554403
12           9715.0          10440.643415
13          11319.0          11155.572896
14          10794.0          10935.454946
15          11346.0           9915.151119
16          11856.0          10313.515782
17           9650.0          10132.251722
18          10085.0          10128.760617
19          11165.0          10055.064281
20          11244.0          11144.347870
21          11652.0          10244.801947
22          11183.0          10806.639605
23          11440.0          10580.533541
24          10200.0          10882.555505
25          11294.0          10455.902918
26           7598.0           9819.698346
27          10911.0          11513.234203
28          10930.0          10448.411889
29          11206.0          10986.655161

From the result table, it can be seen that the model is efficient at most points. The deviation of predicted gold value is lower than 5% compared to the real gold value. However, for some points, the model predicts the result with deviation more than 5%. For example, at index 26, the model predicts 9820 compared to the real value 7598. The deviation is nearly 30%. 